Implemented _eval_nseries() for Hyper

dhruvmendiratta6 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #18193 

#### Brief description of what is fixed or changed
In reference to Issue #18193 the function was implemented. Example of
the working code is as follows:

In [1]: from sympy import hyper

In [2]: from sympy.abc import x, y

In [3]: hyper((1, 2),(1, 2), x)._eval_nseries(x, 7, y)
Out[3]: 1 + x + x**2/2 + x**3/6 + x**4/24 + x**5/120 + x**6/720 + O(x**7)

In [4]: hyper((1, 2),(3, 4, 5), x)._eval_nseries(x, 7, y)
Out[4]: 1 + x/30 + x**2/1200 + x**3/63000 + x**4/4233600 + x**5/355622400 + x**6/36578304000 + O(x**7)


#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* functions
   * Added _eval_nseries() functionality to hyper
<!-- END RELEASE NOTES -->

dhruvmendiratta6 committed:
Implemented _eval_nseries() for Hyper

In reference to Issue #18193 the function was implemented. Example of
the working code is as follows:

In [1]: from sympy import hyper

In [2]: from sympy.abc import x, y

In [3]: hyper((1, 2),(1, 2), x)._eval_nseries(x, 7, y)
Out[3]: 1 + x + x**2/2 + x**3/6 + x**4/24 + x**5/120 + x**6/720 + O(x**7)

In [4]: hyper((1, 2),(3, 4, 5), x)._eval_nseries(x, 7, y)
Out[4]: 1 + x/30 + x**2/1200 + x**3/63000 + x**4/4233600 + x**5/355622400 + x**6/36578304000 + O(x**7)

oscarbenjamin commented:
This will need tests

dhruvmendiratta6 commented:
Actually the following test is still failing. I committed it anyway as the code is working and I couldn't figure out what was going wrong.
` File "c:\users\mendiratta\sympy\sympy\functions\special\tests\test_hyper.py", line 347, in test_limits
    O(k**6) # issue 6350
AssertionError
`

jksuom commented:
There a a couple of things that should be changed in the code. In the failing issue, the argument of the function is `k**2` so that is what should appear in the series expansion. Also I think that `self` should be added to the `super()` call. Then it should be possible to fix the failing test.

jksuom commented on sympy/functions/special/hyper.py:
```diff
@@ -220,6 +220,33 @@ def _eval_rewrite_as_Sum(self, ap, bq, z, **kwargs):
+            for b in bq:
+                Rf2 *= RisingFactorial(b, i)
+
+            terms.append(((Rf1 / Rf2) * (x ** i)) / factorial(i))
```
```suggestion
            terms.append(((Rf1/Rf2) * (arg**i)) / factorial(i))
```

jksuom commented on sympy/functions/special/hyper.py:
```diff
@@ -220,6 +220,33 @@ def _eval_rewrite_as_Sum(self, ap, bq, z, **kwargs):
+        from sympy.functions import factorial, RisingFactorial
+        from sympy import Order, Add
+
+        x0 = self.args[2].limit(x, 0)
```
```suggestion
        arg = self.args[2]
        x0 = arg.limit(x, 0)
```

jksuom commented on sympy/functions/special/hyper.py:
```diff
@@ -220,6 +220,33 @@ def _eval_rewrite_as_Sum(self, ap, bq, z, **kwargs):
+        bq = self.args[1]
+
+        if x0 != 0:
+            return super()._eval_nseries(x, n, logx)
```
```suggestion
           return super()._eval_nseries(self, x, n, logx)
```

jksuom replied:
I'm not sure of this, though. The docs say that if the second argument of `super` is omitted, then the method will be unbound but I don't know what will happen if both arguments are omitted. However, there is an example that seems to indicate that the returned method will be bound to `self`. Perhaps that should be tested.

jksuom replied:
Tests seem to show that the method will be bound. So better leave this unchanged.

dhruvmendiratta6 replied:
I think what you are saying is the way it was done in python 2 and in python 3 both the arguments can be left omitted.

jksuom commented:
Actually, `terms` should be computed with expansion of `arg` as a series in `x`, but that is not needed for the existing tests.

dhruvmendiratta6 committed:
Making suggested changes

Co-Authored-By: Kalevi Suominen <jksuom@gmail.com>

dhruvmendiratta6 committed:
Making suggested changes

Co-Authored-By: Kalevi Suominen <jksuom@gmail.com>

dhruvmendiratta6 committed:
Making suggested changes

Co-Authored-By: Kalevi Suominen <jksuom@gmail.com>

dhruvmendiratta6 commented:
After removing `self` argument from `super()._eval_nseries()` call which was added by mistake there are some differences while calling the `_eval_nseries()` of `hyper` and `exp` which are as follows:

```
>>> hyper((1,2),(1,2), x)._eval_nseries(y,7,None)
hyper((1, 2), (1, 2), x) + O(y**7)

>>> exp(x)._eval_nseries(y,7,None)
exp(x)

>>> exp(x**2)._eval_nseries(x,7,None)
1 + x**2 + x**4/2 + x**6/6 + x**8/24 + x**10/120 + x**12/720 + O(x**14)

>>> hyper((1,2),(1,2), x**2)._eval_nseries(x,7,None)
1 + x**2 + x**4/2 + x**6/6 + O(x**7)
```
is the test failing due to the difference in second output. As what I understood was that maximum power of `x` can be of the order of `n` but that is not the case with `exp._eval_nseries()`

dhruvmendiratta6 committed:
Removing self argument in hyper()_eval_nseries()

dhruvmendiratta6 committed:
Added and modified tests

Simplified expression in test_limits() to make it work with new
implementation of _eval_nseries(). Added new tests for _eval_nseries().
And renamed some variables to make readabilty of the code better

jksuom commented:
Thanks, this is ready to be merged.

dhruvmendiratta6 commented:
Is there any other way apart from adding tests (which is already done) to improve the code coverage
